nucypher/utilities/versioning.py

```diff
@@ -0,0 +1,94 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from abc import abstractmethod, ABC
+from typing import Dict
+
+
+class Versioned(ABC):
+
+    _BRAND_LENGTH = 2  # bytes
+    _VERSION_LENGTH = 2
+    _HEADER_SIZE = _BRAND_LENGTH + _VERSION_LENGTH
+
+    class InvalidHeader(ValueError):
+        """Raised when an unexpected or invalid bytes header is encountered during deserialization"""
+
+    @classmethod
+    def from_bytes(cls, data):
+
+        # Parse brand
+        brand = data[:cls._BRAND_LENGTH]
+        if brand != cls._brand():
+            error = f"Incorrect brand. Expected {cls._brand()}, Got {brand}."
+            if not brand.isalpha():
```

Comment:
Is it just a convenience thing? Separating cases when an incorrect (but branded) object is passed vs some random bytestring?

Reply:
Yes, that's right.  I specifically am trying to control he path of execution for any "unbranded" bytes generated by previous versions of the codebase.  I also added an alphanumeric enforcement in `.header()`.  I don't think there are any reliable ways to determine if a blob of bytes is random or an old workorder for example, but at least we can say that it's completely incompatible either way.